src/chttpd/src/chttpd_changes.erl

```diff
@@ -604,31 +605,49 @@ keep_sending_changes(Args, Acc0, FirstRound) ->
-                                    ?MODULE:keep_sending_changes(
-                                        Args#changes_args{limit = NewLimit},
-                                        ChangesAcc#changes_acc{
-                                            db = Db2,
-                                            user_acc = UserAcc4,
-                                            seq = EndSeq,
-                                            prepend = Prepend2,
-                                            timeout = Timeout,
-                                            timeout_fun = TimeoutFun
-                                        },
-                                        false
-                                    );
-                                _Else ->
-                                    end_sending_changes(Callback, UserAcc3, EndSeq)
+                            AccumulatedTime = timer:now_diff(os:timestamp(), T0) div 1000,
+                            Max = changes_duration(),
+                            case AccumulatedTime > Max of
+                                true ->
+                                    end_sending_changes(Callback, UserAcc4, EndSeq);
+                                false ->
+                                    UserCtx = fabric2_db:get_user_ctx(Db),
+                                    DbOptions1 = [{user_ctx, UserCtx} | DbOptions],
+                                    case fabric2_db:open(fabric2_db:name(Db), DbOptions1) of
+                                        {ok, Db2} ->
+                                            ?MODULE:keep_sending_changes(
+                                                Args#changes_args{limit = NewLimit},
+                                                ChangesAcc#changes_acc{
+                                                    db = Db2,
+                                                    user_acc = UserAcc4,
+                                                    seq = EndSeq,
+                                                    prepend = Prepend2,
+                                                    timeout = Timeout,
+                                                    timeout_fun = TimeoutFun
+                                                },
+                                                false,
+                                                T0
+                                            );
+                                        _Else ->
+                                            end_sending_changes(Callback, UserAcc3, EndSeq)
+                                    end
                             end;
                         {stop, UserAcc4} ->
                             end_sending_changes(Callback, UserAcc4, EndSeq)
                     end
             end
     end.
 
+
+changes_duration() ->
+    %% preserving original (3.x) configuration segment;
+    case config:get("fabric", "changes_duration") of
+        undefined ->
+            infinity;
```

Comment:
Could add "infinity" as an explicit config value.

If we don't have it, it impossible to override the value with say `local.ini` when `default.ini` or other higher config layer, already has it set to an explicit value.

Reply:
this is how it works on 3.x. If you want to impose a duration you set a numeric value. If you do not want to impose a duration you delete the entry (config:delete or `changes_duration =`).

Reply:
That makes it impossible to override a numeric entry with a default infinity in the config chain. We used this pattern in a few places and "infinity" as timeout value is standard in erlang as well:

```
src/couch/src/couch_doc.erl:    MaxLen = case config:get("couchdb", "max_document_id_length", "infinity") of
src/couch_replicator/src/couch_replicator_httpc.erl:        Timeout = case config:get("replicator", "request_timeout", "infinity") of
```